File: haskell/armstrong-numbers/src/ArmstrongNumbers.hs
module ArmstrongNumbers (armstrong) where

digs :: Integral a => a -> [a]
digs 0 = []
digs n = n `mod` 10 : digs (n `div` 10)

armstrong :: Integral a => a -> Bool
armstrong n = n == sum powers where
  powers = map (\x -> x^(length digits)) digits
  digits = digs n

File: fsharp/armstrong-numbers/ArmstrongNumbers.fs
module ArmstrongNumbers

let sumDigits n =
    let digits = List.unfold (fun x ->
                    if x = 0 then None
                    else Some(x%10, x/10)) n
    let nDigits = List.length digits
    digits |> List.map (fun x -> float(x)**float(nDigits)) |> List.sum |> int                    

let isArmstrongNumber (number: int): bool = 
    number = (sumDigits number)